### PR TITLE
feat: Read/Only study mounts on AWS Service Catalog based EC2 Windows workspaces

### PR DESCRIPTION
**Pull request by dpandya**
Issue #, if available:

Description of changes:

Adds support for Read/Only study data on EC2 Windows (AWS Service Catalog) based environments by downloading the study data on local EBS volumes. This PR is **PART 1** and has basic implementation that performs the multipart download once and does not automatically sync further data added to the study after the workspace is launched.

Details:

- Created `s3-synchronizer` tool (in `Go`) to synchronize S3 bucket prefix to the specified destination directory
- Added `environment-tools` SDC to build and push tools (such as s3-synchronizer) to environment bootstrap files S3 bucket. These tools are used by environments (workspaces) and are downloaded at instance bootstrap time. Currently, this only pushes the `s3-synchronizer” and that is ONLY used in Service Catalog based EC2 Windows environments.
- Added `serverless-go-build-tools` serverless plugin to package and deploy Go modules
- Added default setting `environmentsBootstrapBucketPrefix` to make environment files prefix settings driven
- EC2 Windows Service Catalog product CloudFormation template changes to create a striped (RAID 0) volume and mount it to **D** drive by joining 8 EBS volumes and using `s3-synchronizer` tool to download the studies data
- Add `Go` runtime to CodeBuild used in CI/CD pipeline
- README adjustments to mention dependency on Go from the build/deployment machine
- Tested with study containing 1.25 GB data. Got loaded in less than 23 seconds. _Setup:_ 8 stripped gp2 EBS volumes 4GB each on m5.xlarge instance (EBS gp2 default IOPS = 3 IOPS/GB)

> :warning:  **Need to re-import the EC2 Windows Workspace Type to get the ReadOnly on EC2 Windows support**
> :warning:  **Need to re-deploy CI/CD pipeline to include Go runtime. The pipeline will fail without that**

Checklist: 

* [x] Have you successfully deployed to an AWS account with your changes? 
* [x] Have you linted your code locally prior to submission?
* [x] Have you successfully run unit tests and manual tests with your changes locally?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,13 +25,14 @@ The solution also includes a Continuous Integration/Continuous Delivery feature:
...
-Before you can build this project, you need to install [pnpm](https://pnpm.js.org/en/). Run the following command:
+Before you can build this project, please install the following prerequisites.
 
+- **Node.Js:** [Node.js v12.x](https://nodejs.org/en/) or later is required. 
+- **PNPM:** Install [pnpm](https://pnpm.js.org/en/) as follows 
 ```bash
-npm install -g pnpm
+$ npm install -g pnpm
 ```
+- **Go:** You also need to install [Go](https://golang.org/doc/install). `Go` is used for creating a multipart S3 downloader tool that is used in AWS Service Catalog EC2 Windows based research environments.
```
[the remark this answers is not in the document]
Comment (dpandya) [reply README.md:35]: This is PART 1 PR, the implementation of making sure new S3 updates are synced back to the windows machine is underway. We can update the docs but will have to delete/update them as soon as PART 2 (dynamic sync is complete)

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml [hunk 2]
```diff
@@ -81,26 +125,18 @@ Resources:
...
           ToPort: 3389
           CidrIp: !Ref AccessFromCIDRBlock
-        - IpProtocol: tcp
-          FromPort: 80
-          ToPort: 80
-          CidrIp: !Ref AccessFromCIDRBlock
-        - IpProtocol: tcp
-          FromPort: 443
-          ToPort: 443
-          CidrIp: !Ref AccessFromCIDRBlock
```
[the remark this answers is not in the document]
Comment (dpandya) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml:91]: No. Windows instances are access over RDP in SWB and not HTTP/HTTPs

[the remark this answers is not in the document]
Comment (jn1119) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml:91]: Got it, resolving this thread

File: README.md [hunk 3]
```diff
@@ -25,13 +25,14 @@ The solution also includes a Continuous Integration/Continuous Delivery feature:
...
-Node.js v12.x or later is required.
-
-Before you can build this project, you need to install [pnpm](https://pnpm.js.org/en/). Run the following command:
+Before you can build this project, please install the following prerequisites.
 
+- **Node.Js:** [Node.js v12.x](https://nodejs.org/en/) or later is required. 
+- **PNPM:** Install [pnpm](https://pnpm.js.org/en/) as follows 
 ```bash
-npm install -g pnpm
+$ npm install -g pnpm
```
[the remark this answers is not in the document]
Comment (dpandya) [reply README.md:33]: No, I copied this part of docs from other place it had dollar sign indicating this as a shell. I can delete it if needed.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jn1119) [reply README.md:35]: Are you planning on merging this change as it is directly into master. If yes, then I think this particular PR should address the documentation piece. You can publish to a different branch and publish everything including part 1 and part 2 together to master later.

**Review by jn1119 — COMMENTED** [review]
Didn't see any unit tests in s3-synchronizer package. Can we add them in?

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jn1119) [reply README.md:33]: I see, I don't mind it. It's just inconsistent with the entire Readme doc. Either we switch to the dollar convention in entirety or we can keep everything as it is. Your call.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpandya) [reply README.md:35]: I was thinking of not merging right away until the read/write is also available but if we need to release this basic version (with its existing limitations) as is then you are right, this surely needs to be documented. I checked the existing docs, we do not have any place where we mention how to access study data on workspaces, will need to add a section about that and add specific details about windows.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dpandya) [reply README.md:33]: Absolutely, added `$` to all other bash code in this README

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpandya) [reply README.md:35]: Added some info about how to access study data in various workspaces in `docs/docs/user_guide/sidebar/common/workspaces/accessing_a_workspace.md` and also added this limitation for EC2 Windows

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+- Go 1.13.7+
+
+## Usage
+
+```bash
+$ s3-synchronizer-darwin-amd64 -h
+Usage of bin/s3-synchronizer-darwin-amd64:
+  -defaultS3Mounts string
+        A JSON string containing information about the default S3 mounts E.g., [{"id":"some-id","bucket":"some-s3-bucket-name","prefix":"some/s3/prefix/path","writeable":false,"kmsKeyId":"some-kms-key-arn"}]
```
[the remark this answers is not in the document]
Comment (dpandya) [reply addons/addon-raas-s3-copy/packages/s3-synchronizer/README.md:22]: Adjusted this README

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+				log.Printf("Increment wg counter")
+			}
+			mountsCh <- config
+		}
+
+		// Add to the currentMounts
+		currentMounts[s] = struct{}{}
+	}
+	//	time.Sleep(time.Duration(pollInterval) * time.Second)
+	//}
```
[the remark this answers is not in the document]
Comment (dpandya) [reply addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go]: Done

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+			if debug == true {
+				log.Printf("Decrement wg counter")
+			}
+			wg.Done() // Decrement wait group counter everytime we receive config from the mount channel and complete processing it
+		}
+	}()
+
+	// TODO: Add support for fetching S3 mounts dynamically using SWB APIs
+	//// Poll for environment info
+	//for {
```
[the remark this answers is not in the document]
Comment (dpandya) [reply addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go]: Done

**Comment by dpandya** [comment]
> Didn't see any unit tests in s3-synchronizer package. Can we add them in?

Done.

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+			if mountConfig.writeable {
+				go func() {
+					err := setupUploadWatcher(sess, mountConfig, debug)
+					if err == nil {
+						wg.Add(1) // Increment wait group counter everytime we spawn upload watcher
+					} else {
+						log.Printf("Error setting up file watcher: " + err.Error())
+					}
+				}()
+			}
```
Comment (jn1119) [inline addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go:48]: I missed seeing this before somehow. Do we need ship upload-watcher and even the initialization here if that is unfinished?

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/fsnotify/fsnotify"
+)
+
+func setupUploadWatcher(sess *session.Session, config *mountConfiguration, debug bool) error {
```
Comment (jn1119) [inline addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go:16]: I am assuming we are adding unit tests for this later, and not merging this in master right away?

**Review by jn1119 — COMMENTED** [review]
LGTM overall. Had just one comment around unused functionality.

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go (hunk 8 above)
Comment (dpandya) [reply addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-upload-file-watcher.go:16]: Yes. Will add unit tests around this when implementing writes in next PR.

File: addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go (hunk 7 above)
Comment (dpandya) [reply addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer.go]: Done. Will add this along with unit tests when adding Read/Write functionality.

**Review by jn1119 — APPROVED** [review]

**Pull request merged**